xml/encoder_test.go

```diff
@@ -0,0 +1,447 @@
+package xml_test
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/awslabs/smithy-go/xml"
+)
+
+func TestEncoder(t *testing.T) {
+	encoder := xml.NewEncoder()
+
+	obj := encoder.RootElement()
+	obj.Key("stringKey").String("stringValue")
+	obj.Key("integerKey").Integer(1024)
+	obj.Key("floatKey").Float(3.14)
+
+	vk, closeFn := obj.Key("foo").NestedElement()
+	vk.Key("byteSlice").String("Zm9vIGJhcg==")
+
+	if closeFn != nil {
+		closeFn()
+	}
+
+	e := []byte(`<stringKey>stringValue</stringKey><integerKey>1024</integerKey><floatKey>3.14</floatKey><foo><byteSlice>Zm9vIGJhcg==</byteSlice></foo>`)
+	if a := encoder.Bytes(); bytes.Compare(e, a) != 0 {
+		t.Errorf("expected %+q, but got %+q", e, a)
+	}
+
+	if a := encoder.String(); string(e) != a {
+		t.Errorf("expected %s, but got %s", e, a)
+	}
+}
+
+func TestEncodeAttribute(t *testing.T) {
+	encoder := xml.NewEncoder()
+	obj := encoder.RootElement()
+
+	obj.Key("payload", func(t *xml.TagMetadata) {
+		t.AttributeName = "attrkey"
+		t.AttributeValue = "value"
+	}).Null()
+
+	expect := `<payload attrkey="value"></payload>`
+
+	if e, a := expect, encoder.String(); e != a {
+		t.Errorf("expect bodies to match, did not.\n,\tExpect:\n%s\n\tActual:\n%s\n", e, a)
+	}
+}
+
+func TestEncodeNamespace(t *testing.T) {
+	encoder := xml.NewEncoder()
+	obj := encoder.RootElement()
+
+	// nested `payload` shape
+	o := obj.Key("payload")
+	n, closeFn1 := o.NestedElement()
+
+	// nested `namespace` shape
+	n1 := n.Key("namespace", func(t *xml.TagMetadata) {
+		t.NamespacePrefix = "prefix"
+		t.NamespaceURI = "https://example.com"
+	})
+
+	// nested `prefixed` shape
+	n2, closeFn2 := n1.NestedElement()
+	n2.Key("prefixed").String("abc")
+
+	// call all close func's
+	closeFn2()
+	closeFn1()
+
+	e := []byte(`<payload><namespace xmlns:prefix="https://example.com"><prefixed>abc</prefixed></namespace></payload>`)
+	verify(t, encoder, e)
+}
+
+func verify(t *testing.T, encoder *xml.Encoder, e []byte) {
+	if a := encoder.Bytes(); bytes.Compare(e, a) != 0 {
+		t.Errorf("expected %+q, but got %+q", e, a)
+	}
+
+	if a := encoder.String(); string(e) != a {
+		t.Errorf("expected %s, but got %s", e, a)
+	}
```

Comment:
Nit probably want to compare `encoder.String()` with `encoder.Bytes()` to be more specific with the intended behavior that their values are equal.